pulp_smash/cli.py

```diff
@@ -0,0 +1,199 @@
+        ])
+        return '{}({})'.format(type(self).__name__, str_kwargs)
+
+    def check_returncode(self):
+        """Raise ``subprocess.CalledProcessError`` if exit code is non-zero."""
+        if self.returncode:
+            if version_info < (3, 5):  # pragma: no cover
+                args = [self.returncode, self.args, self.stdout]
+            else:  # pragma: no cover
+                args = [self.returncode, self.args, self.stdout, self.stderr]
+            raise subprocess.CalledProcessError(*args)
+
+
+class Client(object):  # pylint:disable=too-few-public-methods
+    """A convenience object for working with a CLI.
+
+    This class provides the ability to execute shell commands on either the
+    local system or a remote system. Here is a simple usage example:
+
+    >>> from pulp_smash import cli, config
+    >>> server_config = config.ServerConfig('localhost')
+    >>> client = cli.Client(server_config)
+    >>> response = client.run(('echo', '-n', 'foo'))
+    >>> response.returncode == 0
+    True
+    >>> response.stdout == 'foo'
+    True
+    >>> response.stderr == ''
+    True
+
+    You can customize how commands are executed and how responses are handled
+    by fiddling with the two public instance attributes:
+
+    ``machine``
+        A `Plumbum`_ machine. :meth:`run` delegates all command execution
+        responsibilities to this object.
+    ``response_handler``
+        A callback function. Each time ``machine`` executes a command, the
+        result is handed to this callback, and the callback's return value is
+        handed to the user.
+
+    If ``server_config.cli_transport`` is ``'local'`` or ``'ssh``, set
+    ``machine`` so that commands run locally or over SSH, respectively. If
+    ``server_config.cli_transport`` is ``None``, guess how to set ``machine``
+    by comparing the hostname embedded in ``server_config.base_url`` against
+    the current system's hostname. If they match, set ``machine`` to execute
+    commands locally; and vice versa.
+
+    :param pulp_smash.config.ServerConfig server_config: Information about the
+        system on which commands will be executed.
+    :param response_handler: A callback function. Defaults to
+        :func:`pulp_smash.cli.code_handler`.
+
+    .. _Plumbum: http://plumbum.readthedocs.org/en/latest/index.html
+    """
+
+    def __init__(self, server_config, response_handler=None):
+        """Initialize this object with needed instance attributes."""
+        # How do we make requests?
+        hostname = _get_hostname(server_config.base_url)
+        if server_config.cli_transport is None:
+            transport = 'local' if hostname == socket.gethostname() else 'ssh'
+        else:
+            transport = server_config.cli_transport
+        if transport == 'local':
+            self.machine = plumbum.machines.local
+        else:  # transport == 'ssh'
+            # Get SSH configuration options from an ssh_config file. See
```

Comment:
Yes, it's correct that `SshMachine` will use `~/.ssh/config` if it is present. Consequently, I do need to edit this section. (My understanding of how Plumbum's `SshMachine` handled SSH config files was murky when I wrote this code.)

My train of thought here is that it'd be cool to configure the `SshMachine` a little bit. For example, it could be nice if we could tell the `SshMachine` to never interact with the user. This would prevent sad things like test sessions on a CI server hanging forever due to SSH key unlock prompts just hanging. It could also be nice to make the `SshMachine` look for an alternate config file. This would prevent Pulp Smash from knowing things it shouldn't (it shouldn't know the key file for logging into my home server or VMs at work), it would let me configure Pulp Smash differently from my regular SSH client (maybe I want Pulp Smash to use a different username, key and port).

But -- I don't see an easy way to do this, and we may be meeting users expectations best by just falling back to standard OpenSSH behaviours.